src/test/threadpool_tests.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2012-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/threadpool.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(threadpool_tests)
+
+BOOST_AUTO_TEST_CASE(threadpool_basic)
+{
+    // Test Cases
+    // 1) Submit tasks and verify completion.
+    // 2) Maintain all threads busy except one.
+    // 3) Wait for work to finish.
+    // 4) Wait for result object.
+    // 5) The task throws an exception, catch must be done in the consumer side.
+    // 6) Busy workers, help them by processing tasks from outside.
+
+    // Test case 1, submit tasks and verify completion.
+    {
+        int num_workers = 3;
+        int num_tasks = 50;
+
+        ThreadPool threadPool;
+        threadPool.Start(num_workers);
+        std::atomic<int> counter = 0;
+        for (int i=0; i<num_tasks; i++) {
```

Comment:
I think a test where each task is assigned slightly different data would be nice. Something like:
```c++
+        std::atomic<int> par_sum{0};
+        for (int i = 0; i < num_tasks; i++) {
+            threadPool.Submit([&par_sum,i]() {
+                par_sum += i;
             });
         }
+        int sync_sum{0};
+        for (int i = 0; i < num_tasks; i++) {
+            sync_sum += i;
+        }
```